modules/core/validator.js

```diff
@@ -481,26 +484,19 @@ export function coreValidator(context) {
       return _headPromise;
     }
 
-    _headGraph = currGraph;  // take snapshot
-    const difference = coreDifference(prevGraph, _headGraph);
-
-    // Gather all entities related to this difference..
-    // For created/modified, use the head graph
-    let entityIDs = difference.extantIDs(true);   // created/modified (true = w/relation members)
-    entityIDs = entityIDsToValidate(entityIDs, _headGraph);
-
-    // For modified/deleted, use the previous graph
-    // (e.g. deleting the only highway connected to a road should create a disconnected highway issue)
-    let previousEntityIDs = difference.deleted().concat(difference.modified()).map(entity => entity.id);
-    previousEntityIDs = entityIDsToValidate(previousEntityIDs, prevGraph);
-    previousEntityIDs.forEach(entityIDs.add, entityIDs);   // concat the sets
+    // If we get here, its time to start validating stuff.
+    _headCache.graph = currGraph;  // take snapshot
+    _completeDiff = context.history().difference().complete();
+    const incrementalDiff = coreDifference(prevGraph, currGraph);
+    const entityIDs = Object.keys(incrementalDiff.complete());
```

Comment:
@bhousel Will this gather entire hierarchy to be validated? On touching a node (e.g. that belongs to a way street, that belongs to a Bus route that belongs to a Route master super relation), all of these would be added for validation.
Previous code would just add node and a way, which I think we established is enough.

Reply:
I don't think it gathers full hierarchies - `difference#complete` only recurses down for Multipolygon relations
https://github.com/openstreetmap/iD/blob/develop/modules/core/difference.js#L202

The point of this "complete" diff is to determine "if the user touches something, what needs to be redrawn" and I think we can reuse that same code to determine "what needs to be revalidated" rather than having a separate logic in the validator.  There's a risk that we might be missing some things that the old code caught, but I'm ok with that for now.

Reply:
Oh yes, but perhaps you misunderstand - I was talking about recurse **up**. The code you linked will go 'up' in parenting chain until it reaches the end.
Actually I think re-using this method is a good move, I am just double checking on how will this impact performance (now modifying one node that I provided as example would queue Route master super relation for validation).

Reply:
I just did a quick look and I didn't find any red-flags that this can hurt perf. Still I want to hear your opinion and if needed we can improve this in additional change.